_posts/2018/2018-02-18-avoid-writing-sql-when-using-activerelation.md

```diff
@@ -0,0 +1,74 @@
+---
+title: 'Avoid Writing SQL When Using ActiveRelation'
+description: 'It's fine, but prefer not to if you can.'
+layout: article
+category: ruby
+image:
+  base: '2018/avoid-writing-sql-when-using-activerelation'
+  alt: 'Coding SQL'
+  source: 'https://unsplash.com/photos/fPkvU7RDmCo'
+  credit: 'Caspar Rubin'
+---
+
+ActiveRelation, which powers the searching and querying engine of ActiveRecord, is a powerful and flexible tool.
+
+
+## Instead of...
+
+...writing literal SQL strings inside an ActiveRelation `#where` method.
+
+```ruby
+Person.where("name = #{ params[:name] } and hidden_at is null")
+```
+
+
+## Or...
+
+...writing literal strings and using the Array method to sanitise data.
+
+```ruby
+Person.where('name = ? and hidden_at is null', params[:name])
+```
+
+
+## Use...
+
+...the hash-style syntax.
+
+```ruby
+Person.where(name: params[:name], hidden_at: nil)
+```
+
+
+## Why?
+
+All methods are valid. The first two versions, where you manually write SQL, was the _only_ way before ActiveRelation was merged into Rails (in version 3.0). However, the “hash style” gives you more flexibility and safety.
+
+The first example is **very** dangerous. **Do not** use string interpolation in your queries. It opens you up to SQL injection attacks: where bad people from the Internet can try and run destructive/exciting statements against your database.
+
+The second example is a better way to write “string style” `#where` methods. But you're still left to write SQL yourself.
+
+I much prefer the final example, when you use the “hash style” syntax. This version is shorter, clearer and likely better highlighted in your editor. Code is made to be read more than written.
+
```

Comment:
* The "hash style" version is shorter, clearer, and likely better highlighted...Remember: code is made to be read more than it is written.

(Take out the "I" part...who cares..? :P)